File: 215/215327/215327_01.c
#include <errno.h>
#include <getopt.h>
#include <inttypes.h>
#include <limits.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <gmp.h>

#define DEFAULT_MAX_N 12UL

typedef struct {
    mpz_t s;
    mpz_t l;
    mpz_t c;
} Counts;

typedef struct {
    size_t q;
    size_t n;
    size_t *word;
    Counts *counts;
} Generator;

static void
usage(FILE *stream, const char *program)
{
    fprintf(stream,
            "Usage: %s [OPTIONS] Q [MAX_N]\n"
            "\n"
            "Default output is a two-column b-file-style table for S_q(n).\n"
            "S_q and L_q start with the OEIS convention 0 1; C_q starts at n=1.\n"
            "MAX_N defaults to %lu. Enumeration is exact but exponential.\n"
            "\n"
            "Options:\n"
            "  -s, --sequence NAME  output S, L, or C (default: S)\n"
            "      --check          show S, L, C, the identity RHS, and check status\n"
            "  -h, --help           show this help\n",
            program, DEFAULT_MAX_N);
}

static size_t
parse_positive_size(const char *text, const char *name)
{
    char *end = NULL;
    uintmax_t value;

    if (text[0] == '\0' || text[0] == '-') {
        fprintf(stderr, "%s must be a positive integer\n", name);
        exit(EXIT_FAILURE);
    }
    errno = 0;
    value = strtoumax(text, &end, 10);
    if (errno == ERANGE || end == text || *end != '\0' || value == 0 ||
        value > SIZE_MAX || value > ULONG_MAX) {
        fprintf(stderr, "%s must be a positive integer in the supported range\n",
                name);
        exit(EXIT_FAILURE);
    }
    return (size_t)value;
}

static void *
xcalloc(size_t count, size_t size)
{
    void *pointer;

    if (size != 0 && count > SIZE_MAX / size) {
        fprintf(stderr, "allocation size overflow\n");
        exit(EXIT_FAILURE);
    }
    pointer = calloc(count, size);
    if (pointer == NULL) {
        fprintf(stderr, "memory allocation failed\n");
        exit(EXIT_FAILURE);
    }
    return pointer;
}

static int
letters_are_smooth(size_t a, size_t b)
{
    return a >= b ? a - b <= 1 : b - a <= 1;
}

/*
 * The FKM recursion emits exactly the lexicographically least representative
 * of every q-ary necklace of length n.  At an emitted representative, period
 * is its least period.  The local rejection below imposes smoothness on the
 * linear representative without changing the FKM canonicality test.
 */
static void
generate_necklaces(Generator *generator, size_t position, size_t period)
{
    size_t copied_letter;
    size_t letter;

    if (position > generator->n) {
        if (generator->n % period != 0) {
            return;
        }
        mpz_add_ui(generator->counts->s, generator->counts->s, 1);
        if (period == generator->n) {
            mpz_add_ui(generator->counts->l, generator->counts->l, 1);
            if (letters_are_smooth(generator->word[generator->n],
                                   generator->word[1])) {
                mpz_add_ui(generator->counts->c, generator->counts->c, 1);
            }
        }
        return;
    }

    copied_letter = generator->word[position - period];
    generator->word[position] = copied_letter;
    if (position == 1 ||
        letters_are_smooth(generator->word[position],
                           generator->word[position - 1])) {
        generate_necklaces(generator, position + 1, period);
    }

    if (copied_letter == generator->q - 1) {
        return;
    }
    letter = copied_letter + 1;
    for (;;) {
        generator->word[position] = letter;
        if (position == 1 ||
            letters_are_smooth(generator->word[position],
                               generator->word[position - 1])) {
            generate_necklaces(generator, position + 1, position);
        }
        if (letter == generator->q - 1) {
            break;
        }
        ++letter;
    }
}

static void
count_smooth_objects(size_t q, size_t n, Counts *counts)
{
    Generator generator;

    if (n == SIZE_MAX) {
        fprintf(stderr, "MAX_N is too large\n");
        exit(EXIT_FAILURE);
    }
    generator.q = q;
    generator.n = n;
    generator.word = xcalloc(n + 1, sizeof(*generator.word));
    generator.counts = counts;
    generate_necklaces(&generator, 1, 1);
    free(generator.word);
}

static int
mobius(size_t n)
{
    size_t value = n;
    size_t prime = 2;
    unsigned int prime_factors = 0;

    while (prime <= value / prime) {
        if (value % prime == 0) {
            value /= prime;
            if (value % prime == 0) {
                return 0;
            }
            ++prime_factors;
            while (value % prime == 0) {
                value /= prime;
            }
        }
        prime += prime == 2 ? 1 : 2;
    }
    if (value > 1) {
        ++prime_factors;
    }
    return prime_factors % 2 == 0 ? 1 : -1;
}

/*
 * If P is the current power of the tridiagonal adjacency matrix A, then
 * (P A)[row,column] is the sum of at most three neighboring entries of P.
 * This computes all required traces exactly in O(max_n q^2) GMP additions.
 */
static void
trace_powers(size_t q, size_t max_n, mpz_t *traces)
{
    size_t cells;
    mpz_t *power;
    mpz_t *next;
    size_t index;
    size_t n;
    size_t row;
    size_t column;

    if (q > SIZE_MAX / q) {
        fprintf(stderr, "Q is too large for the trace matrix\n");
        exit(EXIT_FAILURE);
    }
    cells = q * q;
    power = xcalloc(cells, sizeof(*power));
    next = xcalloc(cells, sizeof(*next));
    for (index = 0; index < cells; ++index) {
        mpz_init(power[index]);
        mpz_init(next[index]);
    }
    for (row = 0; row < q; ++row) {
        mpz_set_ui(power[row * q + row], 1);
    }

    for (n = 1; n <= max_n; ++n) {
        for (row = 0; row < q; ++row) {
            for (column = 0; column < q; ++column) {
                mpz_set(next[row * q + column], power[row * q + column]);
                if (column > 0) {
                    mpz_add(next[row * q + column],
                            next[row * q + column],
                            power[row * q + column - 1]);
                }
                if (column + 1 < q) {
                    mpz_add(next[row * q + column],
                            next[row * q + column],
                            power[row * q + column + 1]);
                }
            }
        }
        for (row = 0; row < q; ++row) {
            mpz_add(traces[n], traces[n], next[row * q + row]);
        }
        {
            mpz_t *temporary = power;
            power = next;
            next = temporary;
        }
    }

    for (index = 0; index < cells; ++index) {
        mpz_clear(power[index]);
        mpz_clear(next[index]);
    }
    free(power);
    free(next);
}

static void
cyclic_lyndon_formula(size_t n, mpz_t *traces, mpz_t result)
{
    mpz_t numerator;
    size_t e;

    mpz_init(numerator);
    for (e = 1; e <= n; ++e) {
        int mu;

        if (n % e != 0) {
            continue;
        }
        mu = mobius(e);
        if (mu > 0) {
            mpz_add(numerator, numerator, traces[n / e]);
        } else if (mu < 0) {
            mpz_sub(numerator, numerator, traces[n / e]);
        }
    }
    if (!mpz_divisible_ui_p(numerator, (unsigned long)n)) {
        fprintf(stderr, "internal error: C numerator is not divisible by %zu\n",
                n);
        mpz_clear(numerator);
        exit(EXIT_FAILURE);
    }
    mpz_divexact_ui(result, numerator, (unsigned long)n);
    mpz_clear(numerator);
}

static void
print_value(size_t n, const mpz_t value)
{
    printf("%zu ", n);
    mpz_out_str(stdout, 10, value);
    putchar('\n');
}

static void
flush_stdout(void)
{
    if (fflush(stdout) == EOF) {
        perror("stdout");
        exit(EXIT_FAILURE);
    }
}

int
main(int argc, char **argv)
{
    static const struct option long_options[] = {
        {"sequence", required_argument, NULL, 's'},
        {"check", no_argument, NULL, 1000},
        {"help", no_argument, NULL, 'h'},
        {NULL, 0, NULL, 0}
    };
    const char *program = strrchr(argv[0], '/');
    char sequence = 'S';
    int check = 0;
    int option;
    size_t q;
    size_t max_n;
    Counts *counts;
    mpz_t *traces;
    mpz_t *formula_c;
    mpz_t *right_hand_sides;
    size_t n;

    program = program == NULL ? argv[0] : program + 1;
    while ((option = getopt_long(argc, argv, "s:h", long_options, NULL)) != -1) {
        switch (option) {
        case 's':
            if (strlen(optarg) != 1 ||
                (optarg[0] != 'S' && optarg[0] != 's' &&
                 optarg[0] != 'L' && optarg[0] != 'l' &&
                 optarg[0] != 'C' && optarg[0] != 'c')) {
                fprintf(stderr, "--sequence must be S, L, or C\n");
                return EXIT_FAILURE;
            }
            sequence = optarg[0];
            if (sequence >= 'a' && sequence <= 'z') {
                sequence = (char)(sequence - 'a' + 'A');
            }
            break;
        case 1000:
            check = 1;
            break;
        case 'h':
            usage(stdout, program);
            return EXIT_SUCCESS;
        default:
            usage(stderr, program);
            return EXIT_FAILURE;
        }
    }
    if (argc - optind < 1 || argc - optind > 2) {
        usage(stderr, program);
        return EXIT_FAILURE;
    }

    q = parse_positive_size(argv[optind], "Q");
    max_n = argc - optind == 2
                ? parse_positive_size(argv[optind + 1], "MAX_N")
                : DEFAULT_MAX_N;
    if (max_n == SIZE_MAX) {
        fprintf(stderr, "MAX_N is too large\n");
        return EXIT_FAILURE;
    }

    counts = xcalloc(max_n + 1, sizeof(*counts));
    traces = xcalloc(max_n + 1, sizeof(*traces));
    formula_c = xcalloc(max_n + 1, sizeof(*formula_c));
    right_hand_sides = xcalloc(max_n + 1, sizeof(*right_hand_sides));
    for (n = 0; n <= max_n; ++n) {
        mpz_init(counts[n].s);
        mpz_init(counts[n].l);
        mpz_init(counts[n].c);
        mpz_init(traces[n]);
        mpz_init(formula_c[n]);
        mpz_init(right_hand_sides[n]);
    }

    if (check) {
        printf("# q=%zu\n", q);
        puts("# n S_q(n) L_q(n) C_q(n) RHS check");
    } else if (sequence != 'C') {
        puts("0 1");
    }
    flush_stdout();

    trace_powers(q, max_n, traces);
    for (n = 1; n <= max_n; ++n) {
        size_t d;

        count_smooth_objects(q, n, &counts[n]);
        cyclic_lyndon_formula(n, traces, formula_c[n]);
        for (d = 1; d < n; ++d) {
            if (n % d == 0) {
                mpz_add(right_hand_sides[n], right_hand_sides[n], counts[d].c);
            }
        }
        mpz_add(right_hand_sides[n], right_hand_sides[n], counts[n].l);

        if (mpz_cmp(counts[n].c, formula_c[n]) != 0) {
            gmp_fprintf(stderr,
                        "C check failed at n=%zu: enumeration=%Zd, formula=%Zd\n",
                        n, counts[n].c, formula_c[n]);
            return EXIT_FAILURE;
        }
        if (mpz_cmp(counts[n].s, right_hand_sides[n]) != 0) {
            gmp_fprintf(stderr,
                        "identity failed at n=%zu: S=%Zd, RHS=%Zd\n",
                        n, counts[n].s, right_hand_sides[n]);
            return EXIT_FAILURE;
        }

        if (check) {
            gmp_printf("%zu %Zd %Zd %Zd %Zd OK\n", n, counts[n].s,
                       counts[n].l, counts[n].c, right_hand_sides[n]);
        } else if (sequence == 'S') {
            print_value(n, counts[n].s);
        } else if (sequence == 'L') {
            print_value(n, counts[n].l);
        } else {
            print_value(n, counts[n].c);
        }
        flush_stdout();
    }

    if (check) {
        fprintf(stderr,
                "Verified C_q(n) independently and S_q(n) = L_q(n) + "
                "sum_{d|n, d!=n} C_q(d) for q=%zu, 1<=n<=%zu.\n",
                q, max_n);
    }

    for (n = 0; n <= max_n; ++n) {
        mpz_clear(counts[n].s);
        mpz_clear(counts[n].l);
        mpz_clear(counts[n].c);
        mpz_clear(traces[n]);
        mpz_clear(formula_c[n]);
        mpz_clear(right_hand_sides[n]);
    }
    free(counts);
    free(traces);
    free(formula_c);
    free(right_hand_sides);
    return EXIT_SUCCESS;
}

File: 215/215327/215327_01.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

require "optparse"

# Count smooth necklaces, smooth Lyndon words, and cyclically smooth Lyndon
# words over the ordered alphabet {0, 1, ..., q - 1}.
#
# Usage:
#   ruby 215327_01.rb [OPTIONS] Q [MAX_N]
#
# The default MAX_N is 12.  Enumeration is exact but exponential, so large
# values of Q or MAX_N can take a long time.

DEFAULT_MAX_N = 12

def parse_positive_integer(text, name)
  value = Integer(text, 10)
  abort "#{name} must be a positive integer" unless value.positive?

  value
rescue ArgumentError
  abort "#{name} must be a positive integer"
end

def divisors(n)
  small = []
  large = []
  d = 1
  while d * d <= n
    if (n % d).zero?
      small << d
      large << n / d unless d * d == n
    end
    d += 1
  end
  small + large.reverse
end

def mobius(n)
  value = n
  prime_factors = 0
  prime = 2

  while prime * prime <= value
    if (value % prime).zero?
      value /= prime
      return 0 if (value % prime).zero?

      prime_factors += 1
      value /= prime while (value % prime).zero?
    end
    prime += prime == 2 ? 1 : 2
  end

  prime_factors += 1 if value > 1
  prime_factors.even? ? 1 : -1
end

def matrix_product(left, right)
  size = left.length
  product = Array.new(size) { Array.new(size, 0) }

  size.times do |row|
    size.times do |middle|
      coefficient = left[row][middle]
      next if coefficient.zero?

      size.times do |column|
        product[row][column] += coefficient * right[middle][column]
      end
    end
  end

  product
end

def adjacency_matrix(q)
  Array.new(q) do |row|
    Array.new(q) { |column| (row - column).abs <= 1 ? 1 : 0 }
  end
end

def trace_powers(q, max_n)
  matrix = adjacency_matrix(q)
  power = Array.new(q) { |row| Array.new(q) { |column| row == column ? 1 : 0 } }
  traces = Array.new(max_n + 1, 0)

  1.upto(max_n) do |n|
    power = matrix_product(power, matrix)
    traces[n] = q.times.sum { |index| power[index][index] }
  end

  traces
end

def cyclic_lyndon_count(n, traces)
  numerator = divisors(n).sum do |e|
    mobius(e) * traces[n / e]
  end
  abort "internal error: C numerator is not divisible by #{n}" unless (numerator % n).zero?

  numerator / n
end

# The FKM recursion generates exactly the lexicographically least
# representative of every q-ary necklace of length n. The period parameter is
# the least period at an emitted representative. Rejecting a newly appended
# letter whose difference from its predecessor exceeds 1 restricts the output
# to smooth representatives without changing the canonical-representative
# test performed by the recursion.
def count_smooth_objects(q, n)
  word = Array.new(n + 1, 0)
  smooth_necklaces = 0
  smooth_lyndon = 0
  cyclic_smooth_lyndon = 0

  generate = lambda do |position, period|
    if position > n
      return unless (n % period).zero?

      smooth_necklaces += 1
      if period == n
        smooth_lyndon += 1
        cyclic_smooth_lyndon += 1 if (word[n] - word[1]).abs <= 1
      end
      return
    end

    copied_letter = word[position - period]
    word[position] = copied_letter
    if position == 1 || (word[position] - word[position - 1]).abs <= 1
      generate.call(position + 1, period)
    end

    (copied_letter + 1).upto(q - 1) do |letter|
      word[position] = letter
      next if position > 1 && (word[position] - word[position - 1]).abs > 1

      generate.call(position + 1, position)
    end
  end

  generate.call(1, 1)
  [smooth_necklaces, smooth_lyndon, cyclic_smooth_lyndon]
end

options = { check: false, sequence: "S" }
parser = OptionParser.new do |opts|
  opts.banner = "Usage: ruby #{File.basename($PROGRAM_NAME)} [OPTIONS] Q [MAX_N]"
  opts.separator ""
  opts.separator "Default output is a two-column b-file-style table for S_q(n)."
  opts.separator "S_q and L_q start with the OEIS convention 0 1; C_q starts at n=1."
  opts.separator "MAX_N defaults to #{DEFAULT_MAX_N}. Enumeration is exact but exponential."
  opts.separator ""
  opts.on("-s", "--sequence NAME", "output S, L, or C (default: S)") do |name|
    options[:sequence] = name.upcase
  end
  opts.on("--check", "show S, L, C, the identity RHS, and check status") do
    options[:check] = true
  end
  opts.on("-h", "--help", "show this help") do
    puts opts
    exit
  end
end

begin
  parser.parse!(ARGV)
rescue OptionParser::ParseError => e
  warn e.message
  warn parser
  exit 1
end

unless (1..2).cover?(ARGV.length)
  warn parser
  exit 1
end
unless %w[S L C].include?(options[:sequence])
  abort "--sequence must be S, L, or C"
end

q = parse_positive_integer(ARGV[0], "Q")
max_n = parse_positive_integer(ARGV[1] || DEFAULT_MAX_N.to_s, "MAX_N")

traces = trace_powers(q, max_n)
counts = Array.new(max_n + 1)

1.upto(max_n) do |n|
  counts[n] = count_smooth_objects(q, n)
end

checked_rows = []

1.upto(max_n) do |n|
  smooth_necklaces, smooth_lyndon, cyclic_smooth_lyndon = counts[n]
  proper_divisor_sum = divisors(n).reject { |d| d == n }.sum do |d|
    counts[d][2]
  end
  right_hand_side = smooth_lyndon + proper_divisor_sum
  formula_c = cyclic_lyndon_count(n, traces)

  unless cyclic_smooth_lyndon == formula_c
    abort "C check failed at n=#{n}: enumeration=#{cyclic_smooth_lyndon}, formula=#{formula_c}"
  end
  unless smooth_necklaces == right_hand_side
    abort "identity failed at n=#{n}: S=#{smooth_necklaces}, RHS=#{right_hand_side}"
  end

  checked_rows << [n, smooth_necklaces, smooth_lyndon,
                   cyclic_smooth_lyndon, right_hand_side, "OK"]
end

if options[:check]
  puts "# q=#{q}"
  puts "# n S_q(n) L_q(n) C_q(n) RHS check"
  checked_rows.each { |row| puts row.join(" ") }
  warn "Verified C_q(n) independently and S_q(n) = L_q(n) + " \
       "sum_{d|n, d!=n} C_q(d) for q=#{q}, 1<=n<=#{max_n}."
else
  column = { "S" => 1, "L" => 2, "C" => 3 }.fetch(options[:sequence])
  puts "0 1" unless options[:sequence] == "C"
  checked_rows.each { |row| puts "#{row[0]} #{row[column]}" }
end
